web3/utils/encoding.py

```diff
@@ -209,3 +210,34 @@ def hexstr_if_str(to_type, hexstr_or_primitive):
     else:
         (primitive, hexstr) = (hexstr_or_primitive, None)
     return to_type(primitive, hexstr=hexstr)
+
+
+def decode_rpc_response(response):
+    try:
+        j_dict = json.loads(to_text(response))
+    except json.decoder.JSONDecodeError:
+        msg = 'decode_rpc_response failure for "{}".'.format(repr(response))
+        raise ValueError(msg)
+    return j_dict
+
+
+@curry
+def encode_rpc_request(rpc_id, method, params):
```

Comment:
Right now, this is tightly coupled to the json-rpc request. What if we wanted to do a similar error message when encoding a genesis block into json format for a file? It would be nice to have something like a:

```py
def friendly_json_encode(obj_to_encode):
    ...
```

This could add helpful messages across this project (and others), pretty much anywhere that json encoding is happening.

Instead of using our special knowledge of `params` here, it could just iterate through all dicts and lists, emitting warnings for each offending piece. Something like:

```py
def friendly_json_encode(obj):
    try:
        encoded = json.dumps(obj)
    except TypeError full_exception:
        if hasattr(obj, 'items'):
            item_errors = '; '.join(json_mapping_errors(obj))
            raise TypeError("Could not encode dictionary because of: %s" % item_errors)
        elif eth_utils.is_list_like(obj):
            element_errors = '; '.join(json_list_errors(obj) 
            raise TypeError("Could not encode list because of: %s" % element_errors)
        else: 
            raise full_exception

                 
def json_mapping_errors(mapping):
    for key, val in mapping.items():
        try:
            friendly_json_encode(val)
        except TypeError as exc:
            yield "value at %r because (%s)" % (key, exc)
 

def json_list_errors(iterable):
    for index, element in enumerate(iterable):
        try:
            friendly_json_encode(element)
        except TypeError as exc:
            yield "item at index %d because (%s)" % (index, exc) 
```

(Totally untested)

Reply:
makes sense. let me have a look. hopefully, i'll have some time tomorrow.

Reply:
@carver do want to accommodate  (optional) encoder/decoder hooks ?
```python
def json_encoder(obj, CustomEncoder=None):
    try:
        encode = json.dumps(obj, cls=CustomEncoder)
         ...
# and
def json_decoder(json_string, obj_hook=None):
    try:
        decode = json.loads(json_string, object_hook=obj_hook)
        ...
```

Reply:
Yeah, sounds great :+1:

Reply:
@carver very sorry but i'm out of pocket this week for both 770 and 771. back next week.

Reply:
@carver  just about to re-start this . just looking at the yield approach outlined, it seems to me we'll end up pushing two errors in just about every invalid case. e.g., *{'date': datetime.datetime.utcnow()}* -- one for mapping/sequence types and one for other types leading to the full exception raised. that in turn prevents the full traversal of mapping or sequence objects. just want to make sure that is the intended behavior. is it?

Reply:
The idea is that all of the fields should be traversed to make a message about all the fields that have errors. If that's not what is happening, then it's a bug. But I think it works.